File: remote/deployments/rest-api-rs/src/pg_contract.rs
//! PG contract surface for `dd-remote-rest-api`.
//!
//! This module is the single import site for the generated `dd_pg_defs`
//! crate (driven by `remote/libs/pg-defs/schema/schema.sql`). Reads /
//! writes against the shared RDS Postgres schema should flow through here
//! so the source-of-truth contract stays single and any future schema
//! change is caught at process startup by the assertions below — long
//! before a stale read or write reaches production.
//!
//! Three layers of wiring:
//!   1. Re-exports: convenient `pg_contract::APP_CONFIG_TABLE`, etc.,
//!      access without pulling `dd_pg_defs` into every callsite.
//!   2. Local column lists: the columns this service actually reads /
//!      writes for each shared table. These are intentionally a subset
//!      of the canonical column set; the runtime assertion below verifies
//!      the subset relationship still holds.
//!   3. Startup assertion: called once from `main()` so a schema regen
//!      that drops a column we depend on fails fast, with a clear error
//!      naming the offending column + table.

// These re-exports are intentionally the full canonical surface for every
// table this service touches. Some are unused today (the API still inlines
// SQL for several tables) but they are explicitly part of the contract
// surface and will be exercised as inline SQL gets migrated to the
// canonical constants. Suppressing the unused-import lint here keeps the
// contract intent visible without a sea of yellow warnings.
#![allow(dead_code, unused_imports)]

pub use dd_pg_defs::{
    AgentContextBlobsRow, AgentContextBlobsStatus, AgentContextEmbeddingsRow,
    AgentRemoteDevArtifactRow, AgentRemoteDevEventRow, AgentRemoteDevTaskRow,
    AgentRemoteDevTaskStatus, AgentRemoteDevThreadRow, AppConfigRow, AppConfigStatus,
    ContainerPoolConfigsRow, KnownGitRepoRow, KnownGitRepoStatus,
    LambdaFunctionContainerBuildStatus, LambdaFunctionRow, LambdaFunctionStatus,
    AGENT_CONTEXT_BLOBS_COLUMNS, AGENT_CONTEXT_BLOBS_TABLE, AGENT_CONTEXT_EMBEDDINGS_COLUMNS,
    AGENT_CONTEXT_EMBEDDINGS_TABLE, AGENT_REMOTE_DEV_ARTIFACTS_COLUMNS,
    AGENT_REMOTE_DEV_ARTIFACTS_TABLE, AGENT_REMOTE_DEV_EVENTS_COLUMNS,
    AGENT_REMOTE_DEV_EVENTS_TABLE, AGENT_REMOTE_DEV_TASKS_COLUMNS, AGENT_REMOTE_DEV_TASKS_TABLE,
    AGENT_REMOTE_DEV_THREADS_COLUMNS, AGENT_REMOTE_DEV_THREADS_TABLE, APP_CONFIG_COLUMNS,
    APP_CONFIG_TABLE, KNOWN_GIT_REPOS_COLUMNS, KNOWN_GIT_REPOS_TABLE, LAMBDA_FUNCTIONS_COLUMNS,
    LAMBDA_FUNCTIONS_TABLE,
};

/// Columns the local SELECT in `lambda_select_sql()` returns and that
/// `row_to_lambda_function` reads by name. These must remain a strict
/// subset of `LAMBDA_FUNCTIONS_COLUMNS`; if `schema.sql` ever drops one
/// of these, the startup assertion below fires immediately.
pub const LOCAL_LAMBDA_FUNCTIONS_READ_COLUMNS: &[&str] = &[
    "id",
    "slug",
    "display_name",
    "description",
    "runtime",
    "entry_command",
    "function_body",
    "reuse_key",
    "idle_timeout_seconds",
    "max_run_ms",
    "containerized",
    "container_image",
    "container_build_status",
    "container_build_error",
    "container_built_at",
    "status",
    "labels",
    "meta_data",
    "last_invoked_at",
    "created_at",
    "updated_at",
];

/// Tables this service writes to via INSERT / UPDATE / DELETE. Each entry
/// must match a canonical table constant; the startup assertion proves
/// the table name still exists in the canonical schema.
pub const LOCAL_WRITABLE_TABLES: &[&str] = &[
    AGENT_REMOTE_DEV_EVENTS_TABLE,
    AGENT_REMOTE_DEV_TASKS_TABLE,
    AGENT_REMOTE_DEV_THREADS_TABLE,
    AGENT_CONTEXT_BLOBS_TABLE,
    AGENT_CONTEXT_EMBEDDINGS_TABLE,
    KNOWN_GIT_REPOS_TABLE,
    LAMBDA_FUNCTIONS_TABLE,
];

/// Tables this service reads from. Same shape as `LOCAL_WRITABLE_TABLES`
/// — every entry has to resolve to a canonical constant.
pub const LOCAL_READABLE_TABLES: &[&str] = &[
    AGENT_REMOTE_DEV_EVENTS_TABLE,
    AGENT_REMOTE_DEV_TASKS_TABLE,
    AGENT_REMOTE_DEV_THREADS_TABLE,
    AGENT_CONTEXT_BLOBS_TABLE,
    AGENT_CONTEXT_EMBEDDINGS_TABLE,
    KNOWN_GIT_REPOS_TABLE,
    LAMBDA_FUNCTIONS_TABLE,
];

fn assert_columns_subset(local: &[&str], canonical: &[&str], table: &str) {
    for &column in local {
        assert!(
            canonical.contains(&column),
            "dd-remote-rest-api expected column `{column}` on table `{table}`, but the canonical \
             schema in remote/libs/pg-defs/schema/schema.sql no longer includes it. Either restore \
             the column or stop reading it from {}.",
            std::module_path!()
        );
    }
}

fn assert_table_in_canonical_set(table: &str, canonical_tables: &[&str]) {
    assert!(
        canonical_tables.contains(&table),
        "dd-remote-rest-api expected `{table}` to be one of the canonical schema tables \
         (remote/libs/pg-defs/schema/schema.sql), but it isn't. Either restore the table \
         or update {} to stop referencing it.",
        std::module_path!()
    );
}

/// Call once from `main()` before binding the HTTP listener. Panics with
/// a clear message naming the missing column / table if `schema.sql`
/// has drifted away from what this service reads or writes.
pub fn assert_canonical_schema_matches_local_reads() {
    assert_columns_subset(
        LOCAL_LAMBDA_FUNCTIONS_READ_COLUMNS,
        LAMBDA_FUNCTIONS_COLUMNS,
        LAMBDA_FUNCTIONS_TABLE,
    );

    // The other tables don't have a strict-subset column contract yet
    // (they're served by inline SQL today). For now we lock in the
    // table-name surface so a rename in schema.sql trips us at startup.
    let canonical_tables: &[&str] = &[
        APP_CONFIG_TABLE,
        AGENT_REMOTE_DEV_ARTIFACTS_TABLE,
        AGENT_REMOTE_DEV_EVENTS_TABLE,
        AGENT_REMOTE_DEV_TASKS_TABLE,
        AGENT_REMOTE_DEV_THREADS_TABLE,
        AGENT_CONTEXT_BLOBS_TABLE,
        AGENT_CONTEXT_EMBEDDINGS_TABLE,
        KNOWN_GIT_REPOS_TABLE,
        LAMBDA_FUNCTIONS_TABLE,
    ];
    for &table in LOCAL_READABLE_TABLES {
        assert_table_in_canonical_set(table, canonical_tables);
    }
    for &table in LOCAL_WRITABLE_TABLES {
        assert_table_in_canonical_set(table, canonical_tables);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn local_lambda_columns_are_subset_of_canonical() {
        assert_canonical_schema_matches_local_reads();
    }

    #[test]
    fn lambda_functions_table_name_matches_canonical() {
        assert_eq!(LAMBDA_FUNCTIONS_TABLE, "lambda_functions");
    }

    #[test]
    fn app_config_status_round_trips() {
        assert_eq!(AppConfigStatus::Active.as_str(), "active");
        assert_eq!(AppConfigStatus::Paused.as_str(), "paused");
        assert_eq!(AppConfigStatus::Archived.as_str(), "archived");
        assert!(AppConfigStatus::try_from("nope").is_err());
    }
}

File: remote/deployments/dart-server/lib/db/pg_contract.dart
/// Single import site for the canonical pg-defs contract surface.
///
/// This module mirrors the role of [`rest-api-rs/src/pg_contract.rs`](../../../../rest-api-rs/src/pg_contract.rs)
/// for dd-dart-server. Reads / writes against the shared RDS Postgres
/// schema flow through here so the source-of-truth contract stays
/// single, and any future schema regen is caught at process startup
/// long before a stale read or write reaches production.
///
/// Three layers:
///
///   1. **Re-exports** — convenient `pg_contract.appConfigTable`, etc.
///      access without pulling `dd_pg_defs` into every callsite.
///   2. **Local table lists** — the tables this service actually reads
///      from / writes to. These get checked against the canonical
///      constants at startup.
///   3. **Startup assertion** — [assertPgContract] is called once from
///      `main()` so a schema regen that drops a table we depend on
///      fails fast, with a clear error naming the offending table.
///
/// We deliberately do **not** generate SQL or table DDL here. The schema
/// authority is `remote/libs/pg-defs/schema/schema.sql`; we are an
/// adapter only. See `AGENTS.md` for the full rule.
library;

import 'package:dd_pg_defs/pg_defs.dart' as pg;

// ---------------------------------------------------------------------------
// 1. Re-exports — table-name + select-SQL constants and Row classes for
//    every pg-defs table this service is allowed to consume. Keep this
//    block in alphabetical order so audits stay readable.
// ---------------------------------------------------------------------------

const appConfigTable = pg.appConfigTable;
const appConfigSelectSql = pg.appConfigSelectSql;
const appConfigStatusValues = pg.appConfigStatusValues;
typedef AppConfigRow = pg.AppConfigRow;

const containerPoolConfigsTable = pg.containerPoolConfigsTable;
const containerPoolConfigsSelectSql = pg.containerPoolConfigsSelectSql;
const containerPoolConfigsStatusValues = pg.containerPoolConfigsStatusValues;
typedef ContainerPoolConfigsRow = pg.ContainerPoolConfigsRow;

const knownGitRepoTable = pg.knownGitRepoTable;
const knownGitRepoSelectSql = pg.knownGitRepoSelectSql;
typedef KnownGitRepoRow = pg.KnownGitRepoRow;

const presenceConvsTable = pg.presenceConvsTable;
const presenceConvsSelectSql = pg.presenceConvsSelectSql;
const presenceConvsStatusValues = pg.presenceConvsStatusValues;
typedef PresenceConvsRow = pg.PresenceConvsRow;

const presenceConvMembersTable = pg.presenceConvMembersTable;
const presenceConvMembersSelectSql = pg.presenceConvMembersSelectSql;
const presenceConvMembersRoleValues = pg.presenceConvMembersRoleValues;
const presenceConvMembersStatusValues = pg.presenceConvMembersStatusValues;
typedef PresenceConvMembersRow = pg.PresenceConvMembersRow;

const presenceUsersTable = pg.presenceUsersTable;
const presenceUsersSelectSql = pg.presenceUsersSelectSql;
typedef PresenceUsersRow = pg.PresenceUsersRow;

const presenceEventsTable = pg.presenceEventsTable;
const presenceEventsSelectSql = pg.presenceEventsSelectSql;
const presenceEventsOpValues = pg.presenceEventsOpValues;
typedef PresenceEventsRow = pg.PresenceEventsRow;

const presenceConsumerCheckpointsTable = pg.presenceConsumerCheckpointsTable;
const presenceConsumerCheckpointsSelectSql =
    pg.presenceConsumerCheckpointsSelectSql;
typedef PresenceConsumerCheckpointsRow = pg.PresenceConsumerCheckpointsRow;

// ---------------------------------------------------------------------------
// 2. Local table lists — what this service touches. Both lists must
//    only contain canonical table names that exist in pg-defs; the
//    startup assertion below proves that.
// ---------------------------------------------------------------------------

/// Tables this service reads from (SELECTs).
const localReadableTables = <String>[
  presenceConvsTable,
  presenceConvMembersTable,
  presenceUsersTable,
  presenceEventsTable,
  presenceConsumerCheckpointsTable,
  appConfigTable,
];

/// Tables this service writes to (INSERT / UPDATE / DELETE).
///
/// The `presence_events` table is write-only here for the outbox: every
/// presence/conversation mutation that should fan out across pods rides
/// through the canonical insert + LISTEN/NOTIFY pipeline declared in
/// `schema.sql`. We do not write to `presence_*` data tables directly
/// from individual session isolates — the supervisor on the main isolate
/// owns those writes (and those will land in a follow-up PR; this
/// contract module is the surface they will use).
const localWritableTables = <String>[
  presenceConvsTable,
  presenceConvMembersTable,
  presenceUsersTable,
  presenceEventsTable,
  presenceConsumerCheckpointsTable,
];

/// Every canonical pg-defs table this service knows about (read or
/// write). Used by [assertPgContract] to verify the bindings are still
/// in sync with the canonical schema after a regen.
final allReferencedTables = <String>{
  ...localReadableTables,
  ...localWritableTables,
};

/// Snapshot of every canonical table we currently re-export. Used as a
/// belt-and-braces check against a regen that renames a table.
const _exportedTables = <String>[
  appConfigTable,
  containerPoolConfigsTable,
  knownGitRepoTable,
  presenceConvsTable,
  presenceConvMembersTable,
  presenceUsersTable,
  presenceEventsTable,
  presenceConsumerCheckpointsTable,
];

// ---------------------------------------------------------------------------
// 3. Startup assertion.
// ---------------------------------------------------------------------------

/// Verify our local table-name lists still match the canonical pg-defs
/// constants. Throws if a referenced table does not exist in the
/// generated adapter — that means the schema was regenerated, the table
/// was renamed/dropped, and our wiring needs to follow.
///
/// Call this exactly once from `main()`, ideally before binding the
/// HTTP server, so a regen mismatch fails fast and loudly rather than
/// surfacing as a runtime SQL error against a live RDS hours later.
void assertPgContract() {
  final canonicalTables = <String>{
    pg.appConfigTable,
    pg.containerPoolConfigsTable,
    pg.knownGitRepoTable,
    pg.agentContextBlobsTable,
    pg.agentContextEmbeddingsTable,
    pg.agentRemoteDevThreadTable,
    pg.agentRemoteDevTaskTable,
    pg.agentRemoteDevEventTable,
    pg.agentRemoteDevBreadcrumbTable,
    pg.agentRemoteDevArtifactTable,
    pg.agentRemoteDevRuntimeLockTable,
    pg.lambdaFunctionTable,
    pg.containerPoolImageRevisionsTable,
    pg.containerPoolBuildRunsTable,
    pg.presenceConvsTable,
    pg.presenceConvMembersTable,
    pg.presenceUsersTable,
    pg.presenceEventsTable,
    pg.presenceConsumerCheckpointsTable,
  };

  final missingExports = _exportedTables
      .where((t) => !canonicalTables.contains(t))
      .toList(growable: false);
  if (missingExports.isNotEmpty) {
    throw StateError(
      'pg_contract: re-exported tables no longer exist in dd_pg_defs: '
      '${missingExports.join(', ')}. '
      'Schema regen detected; update pg_contract.dart to follow.',
    );
  }

  final missingReferences = allReferencedTables
      .where((t) => !canonicalTables.contains(t))
      .toList(growable: false);
  if (missingReferences.isNotEmpty) {
    throw StateError(
      'pg_contract: localReadableTables / localWritableTables reference '
      'tables not present in dd_pg_defs: ${missingReferences.join(', ')}. '
      'Either regenerate pg-defs to restore them, or remove them from '
      'this service\'s contract surface.',
    );
  }
}

/// Returns a snapshot suitable for logging at boot or surfacing on
/// `/dart/admin/db`. Pure data — no side effects.
Map<String, Object?> pgContractSnapshot() => <String, Object?>{
      'exported': _exportedTables,
      'readable': localReadableTables,
      'writable': localWritableTables,
    };
